### ISSUE TITLE
Cosine similarity Op request

### ISSUE THREAD

**Issue by vraintanakarei** [onnx/onnx#6586]
# CosineSimilarity

### Describe the operator

I propose to add the cosine similarity to the ONNX operator set.
Cosine similarity is widely used in LLM, metric learning, and many other areas and it is already implemented in [PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html), [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/keras/losses/cosine_similarity) and [scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html).
[Cosine similarity - Wikipedia](https://en.wikipedia.org/wiki/Cosine_similarity)

### Can this operator be constructed using existing onnx operators?

Cosine similarity operators can be constructed in the following two ways using existing onnx operators, but these processes require huge memory to store intermediate results, especially when the input tensor is large.
Also, following the first equation may result in inner products overflow particularly when using low precision (such as FP16).

``` math
\begin{align}
\mathrm{CosineSimilarity}(x_1, x_2)
&= \frac{x_1 \cdot x_2}{\|x_1\|_2 \cdot \|x_2\|_2} & \cdots & (1) \\
&= \frac{x_1}{\|x_1\|_2} \cdot \frac{x_2}{\|x_2\|_2} & \cdots & (2)
\end{align}
```

I believe that a specialized implementation would allow for greater memory efficiency and computational stability.

### Is this operator used by any model currently? Which one?
I don't know.

### Are you willing to contribute it? (Y/N)
N
